src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -944,23 +944,68 @@ namespace cryptonote
     return ret;
   }
   //-----------------------------------------------------------------------------------------------
-  bool core::handle_incoming_txs(const std::vector<tx_blob_entry>& tx_blobs, std::vector<tx_verification_context>& tvc, bool keeped_by_block, bool relayed, bool do_not_relay)
+  bool core::handle_incoming_txs(std::vector<tx_blob_entry> tx_blobs, std::vector<tx_verification_context>& tvc, bool allow_defering, bool keeped_by_block, bool relayed, bool do_not_relay)
   {
     TRY_ENTRY();
     CRITICAL_REGION_LOCAL(m_incoming_tx_lock);
 
-    struct result { bool res; cryptonote::transaction tx; crypto::hash hash; };
-    std::vector<result> results(tx_blobs.size());
+    const size_t prior_txs = m_defered_txs.size();
+
+    for (size_t i = 0; i < tx_blobs.size(); ++i)
+      m_defered_txs.push_back({std::move(tx_blobs[i]), keeped_by_block, relayed, do_not_relay});
+
+    const bool defer = allow_defering && !keeped_by_block && m_defered_txs.size() < DEFER_TX_THRESHOLD;
+    if (defer)
+    {
+      MDEBUG("defering tx verification");
+      return true;
+    }
 
+    std::vector<tx_verification_context> tvc_all;
+    if (!verify_defered_txes(tvc_all))
+      return false;
+    CHECK_AND_ASSERT_MES(tvc_all.size() == tx_blobs.size() + prior_txs, false, "Unexpected number of returned tvc");
     tvc.resize(tx_blobs.size());
+    for (size_t i = 0; i < tvc.size(); ++i)
+      tvc[i] = tvc_all[i + prior_txs];
+    return true;
+    CATCH_ENTRY_L0("core::handle_incoming_txs()", false);
+  }
+  //-----------------------------------------------------------------------------------------------
+  bool core::verify_defered_txes(std::vector<tx_verification_context> &tvc)
+  {
+    PERF_TIMER(verify_defered_txes);
+    TRY_ENTRY();
+    CRITICAL_REGION_LOCAL(m_incoming_tx_lock);
+
+    if (m_defered_txs.empty())
+      return true;
+
+    struct result { bool res; cryptonote::transaction tx; crypto::hash hash; };
+    std::vector<result> results(m_defered_txs.size());
+    std::vector<int> dupes(m_defered_txs.size(), -1);
+
+    tvc.resize(m_defered_txs.size());
     tools::threadpool& tpool = tools::threadpool::getInstance();
     tools::threadpool::waiter waiter;
-    std::vector<tx_blob_entry>::const_iterator it = tx_blobs.begin();
-    for (size_t i = 0; i < tx_blobs.size(); i++, ++it) {
+    std::vector<defered_tx_t>::const_iterator it = m_defered_txs.begin();
+    for (size_t i = 0; i < m_defered_txs.size(); i++, ++it)
+    {
+      // don't process dupes
+      for (size_t j = 0; j < i; ++j)
+      {
+        if (m_defered_txs[j].blob.blob == m_defered_txs[i].blob.blob)
+        {
+          dupes[i] = (int)j;
+          break;
+        }
+      }
+      if (dupes[i] >= 0)
+        continue;
```

Comment:
This makes the scanning cost quadratic in the number of txes being batched. How about using `std::unordered_map` like this?

```diff
--- a/src/cryptonote_core/cryptonote_core.cpp
+++ b/src/cryptonote_core/cryptonote_core.cpp
@@ -984,6 +984,7 @@ namespace cryptonote
     struct result { bool res; cryptonote::transaction tx; crypto::hash hash; };
     std::vector<result> results(m_defered_txs.size());
     std::vector<int> dupes(m_defered_txs.size(), -1);
+    std::unordered_map<blobdata, int> unique_index;
 
     tvc.resize(m_defered_txs.size());
     tools::threadpool& tpool = tools::threadpool::getInstance();
@@ -992,16 +993,16 @@ namespace cryptonote
     for (size_t i = 0; i < m_defered_txs.size(); i++, ++it)
     {
       // don't process dupes
-      for (size_t j = 0; j < i; ++j)
+      auto found = unique_index.find(m_defered_txs[i].blob.blob);
+      if (found != unique_index.end())
       {
-        if (m_defered_txs[j].blob.blob == m_defered_txs[i].blob.blob)
-        {
-          dupes[i] = (int)j;
-          break;
-        }
-      }
-      if (dupes[i] >= 0)
+        dupes[i] = found->second;
         continue;
+      }
+      else
+      {
+        unique_index.insert({m_defered_txs[i].blob.blob, unique_index.size()});
+      }
       tpool.submit(&waiter, [&, i, it] {
         try
         {
```